### PR TITLE
Promisify features

### PR DESCRIPTION
**Pull request by alexdavid**
@kevgo @charlierudolph This promisifies the step definitions using generators

### CODE REVIEW

File: test/feature/support/helpers.coffee [hunk 1]
```diff
@@ -1,8 +1,8 @@
-async = require 'async'
-fs = require 'fs'
-fsExtra = require 'fs-extra'
+Promise = require 'bluebird'
+async = Promise.coroutine
```
Comment (charlierudolph) [inline test/feature/support/helpers.coffee:2]: nice idea to get the closer to the async / await syntax.

File: test/feature/support/helpers.coffee [hunk 2]
```diff
@@ -52,43 +52,42 @@ class Helpers
...
       port: @port
       pathname: urlPath
 
-    request {method, uri}, done
+    yield request {method, uri, resolveWithFullResponse: yes, simple: false}
 
 
   # Starts up server.coffee in @appPath
-  startApp: (done) ->
+  startApp: -> new Promise (resolve) =>
```
Comment (charlierudolph) [inline test/feature/support/helpers.coffee]: Looking at this function seems odd to me now that this will never reject. I almost feel it should have its own timeout instead of letting the cucumber step timeout so it can give a more useful error. Can't be dealt with in a separate PR though.

**Comment by charlierudolph** [comment]
Looks awesome. I'm thinking cucumber-js should actually move to using `Bluebird.coroutine` to run generators but open up the API so a user can provide a function to run generators if they want to use something else.

File: test/feature/support/helpers.coffee (hunk 2 above)
Comment (alexdavid) [inline test/feature/support/helpers.coffee]: 👍

File: circle.yml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 machine:
   node:
-    version: 5.1.0
+    version: 6.2.1
```
Comment (kevgo) [inline circle.yml:3]: Version 6.1 is preinstalled on the "Trusty" image, so your tests run faster if you stick with that version. You have to change the image in the Circle settings.

File: package.json [hunk 4]
```diff
@@ -15,7 +15,7 @@
     "lodash": "^4.4.0"
   },
   "devDependencies": {
-    "async": "^1.4.0",
+    "bluebird": "^3.4.1",
```
Comment (kevgo) [inline package.json:18]: What do you think about moving to fixed versions (without `^`)? It makes `david` work much better, and causes less breakage from packages not following semver.

File: test/feature/step_definitions/eval_definitions.coffee [hunk 5]
```diff
@@ -3,24 +3,18 @@
...
 
-  @Then /^the routing helper "([^"]+)" is undefined$/, (helper, done) ->
-    @makeRequest 'get', "/eval/return String(#{helper})", (err, _, responseBody) ->
-      return done err if err
-      expect(responseBody).to.equal 'undefined'
-      done()
+
+  @Then /^the routing helper "([^"]+)" is undefined$/, (helper) ->
+    response = yield @makeRequest 'get', "/eval/return String(#{helper})"
+    expect(response.body).to.equal 'undefined'
```
Comment (kevgo) [inline test/feature/step_definitions/eval_definitions.coffee:20]: `.to.be.undefined`. Or are you expecting a string "undefined" here?

File: test/feature/support/before_after.coffee [hunk 6]
```diff
@@ -1,22 +1,21 @@
-portfinder = require 'portfinder'
-tmp = require 'tmp'
+Promise = require 'bluebird'
+portfinder = Promise.promisifyAll require 'portfinder'
```
Comment (kevgo) [inline test/feature/support/before_after.coffee:2]: Does this work?

``` coffeescript
{promisifyAll} = require 'bluebird'
portfinder = promisifyAll require 'portfinder'
```

Comment (kevgo) [inline test/feature/support/before_after.coffee:2]: Or:

``` coffeescript
P = require('bluebird).promisifyAll
portfinder = P require 'portfinder'
tmp = P require 'tmp'
```

**Comment by kevgo** [comment]
Cool, looks good! And great to see a side-by-side comparison of the two approaches. Some observations:
- using Nitroglycerin would result in comparable vertical space savings. Repeated code like `return done err if err` should be dried up for a fair comparison of solutions.
- most of your async helper methods trade two `done` for one `async` and `yield`, which is actually more text to type
- the `yield` feels out of place. Yield to me means the method stops here and defers control to the caller with another entry of the result stream. In this case that's not correct. The caller of your async helper methods (a cucumber step implementation) isn't called, but some invisible intermediate code created by `async`. This means `yield` now has two meanings in JS: as a hack for making async code look synchronous, and for processing of large (or infinite) collections where the next iteration is only computed when actually needed. The ES7 syntax should improve this.

File: package.json (hunk 4 above)
Comment (alexdavid) [inline package.json:18]: Yeah, I'll do this in a separate PR

File: test/feature/step_definitions/eval_definitions.coffee (hunk 5 above)
Comment (alexdavid) [inline test/feature/step_definitions/eval_definitions.coffee:20]: I'm expecting a string here because of the way this has to work unfortunately. To get fully isolated exprestive tests I call child_process.fork, but that means I had to create an `eval` route to inspect values of things like the routing helpers. I suppose I could create a higher level RPC wrapper around `@makeRequest 'eval/...'` that passes the response to JSON.parse.

File: test/feature/support/before_after.coffee (hunk 6 above)
Comment (alexdavid) [inline test/feature/support/before_after.coffee:2]: Yeah this could work, but would make it harder to change if I ever need another bluebird function or the bluebird `Promise` constructor.

I don't feel too strongly either way about it though

Comment (kevgo) [inline test/feature/support/before_after.coffee:2]: I don't feel strongly either, just thinking out loud. Its a pretty common pattern with certain libraries like prelude-ls. Not so much with others  :)

File: test/feature/step_definitions/eval_definitions.coffee (hunk 5 above)
Comment (kevgo) [inline test/feature/step_definitions/eval_definitions.coffee:20]: I think comparing output of calls to subshells as string is good enough here. YAGNI.
